internal/controller/management/serviceaccounts/serviceaccounts.go

```diff
@@ -0,0 +1,301 @@
+		}
+		logger.Error(err, "Failed to get ServiceAccount")
+		return ctrl.Result{}, err
+	}
+
+	// Handle ServiceAccount deletion or updates where the
+	// controller label has been removed from the ServiceAccount.
+	// This indicates that the ServiceAccount is no longer managed by the controller,
+	// and we need to clean up any associated RoleBindings.
+	if sa.DeletionTimestamp != nil || !hasControllerLabel(sa) {
+		logger.Debug("Deleting RoleBindings for ServiceAccount", "serviceaccount", sa.Name)
+		if err := r.removeControllerPermissionsFn(ctx, req.NamespacedName); err != nil {
+			return ctrl.Result{}, err
+		}
+		controllerutil.RemoveFinalizer(sa, kargoapi.FinalizerName)
+		if err := r.updateServiceAccountFn(ctx, sa); err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.Debug("Removed finalizer from ServiceAccount", "serviceaccount", sa.Name)
+		return ctrl.Result{}, nil
+	}
+
+	// If we get to here, we had a not found error and we can proceed with
+	// creating the RoleBindings.
+
+	// Add a finalizer to the ServiceAccount to prevent premature deletion
+	// before RoleBindings are removed. The AddFinalizer function:
+	// - Returns false if the finalizer is already present, avoiding redundant logs.
+	// - Adds the finalizer if not present, ensuring proper cleanup.
+	if controllerutil.AddFinalizer(sa, kargoapi.FinalizerName) {
+		if err := r.updateServiceAccountFn(ctx, sa); err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.Debug("Added finalizer to ServiceAccount", "serviceaccount", sa.Name)
+		return ctrl.Result{}, nil
+	}
+
+	logger.Debug("Creating RoleBindings for ServiceAccount",
+		"serviceaccount", sa.Name,
+	)
+	return ctrl.Result{}, r.ensureControllerPermissionsFn(ctx, req.NamespacedName)
+}
+
+// ensureControllerPermissions grants the controller ServiceAccount necessary access to all Projects
+// by creating RoleBindings in each Project's namespace. This function ensures that the
+// controller has the necessary permissions to manage resources across all Projects.
+func (r *reconciler) ensureControllerPermissions(ctx context.Context, sa types.NamespacedName) error {
+
+	roleBindingName := fmt.Sprintf("%s-readonly-secrets", sa.Name)
+
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"serviceaccount", sa.Name,
+		"roleBinding", roleBindingName,
+	)
+	logger.Debug("starting to create RoleBindings for all Projects")
+
+	projectList := &kargoapi.ProjectList{}
+	if err := r.listProjectFn(ctx, projectList); err != nil {
+		return fmt.Errorf("error listing Projects: %w", err)
+	}
+
+	// Loop through each Project to create or update the corresponding RoleBinding.
+	for _, project := range projectList.Items {
+
+		roleBinding := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      roleBindingName,
+				Namespace: project.Name,
+			},
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: rbacv1.GroupName,
+				Kind:     "ClusterRole",
+				Name:     "kargo-controller-secrets-readonly",
```

Comment:
This occurs in enough places that it probably ought to be a constant. This is meant to prevent the scenario where bugs creep over time due this value being updated in some locations and not others.